### DISCUSSION TITLE
how to upgrade to a later docker image?

### DISCUSSION THREAD

**Discussion by diegoquintanav** [br3ndonland/inboard#52] [Q&A] [answered]
I am using `ghcr.io/br3ndonland/inboard:fastapi-0.10.4-python3.9` and I would like to update to a more recent version, e.g. `ghcr.io/br3ndonland/inboard:fastapi-0.23.1-python3.9`. How should I upgrade? Simply replacing the base docker image leads to many errors, the first one that 'inboard' can't be imported from python. I'm guessing that the handling of poetry virtualenvs changes from one version to another.

I need to upgrade because fastapi has a starlette dependency that is incompatible with `jinja2=^3.1.0`.

**Comment by br3ndonland** [comment]
Happy to help with this.

Yes, the Poetry configuration was updated in [inboard 0.14](https://github.com/br3ndonland/inboard/releases/tag/0.14.0). There are some [details on Docker and Poetry in the docs](https://inboard.bws.bio/docker#docker-and-poetry).

I haven't seen any errors when upgrading to inboard 0.14 though. All the applications I run were able to upgrade without any problems.

To help resolve the errors, I will need to learn a little more about your project.

- If it's open-source, can you share a link to the repo?
- If it's closed-source, can you provide an example of your Dockerfile and requirements files (pyproject.toml or requirements.txt)?

Also do you have any error tracebacks? That would help debug the error you're seeing.

**Comment by diegoquintanav** [comment, marked as answer]
Hi, I'm using boilerplate code from another repo, and its dockerfile looks like ![this](https://github.com/whythawk/full-stack-fastapi-postgresql/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/backend/backend.dockerfile) and its `pyproject.toml` looks like ![this](https://github.com/whythawk/full-stack-fastapi-postgresql/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app/pyproject.toml) 

My project is private at this time, but it has not changed that much from the original versions in this aspect (CLI + 3rd library settings e.g. pytest). My Dockerfile looks like this

```Dockerfile
# updated from fastapi-0.10.4-python3.9
FROM ghcr.io/br3ndonland/inboard:fastapi-0.23.1-python3.9

# Copy poetry.lock* in case it doesn't exist in the repo
COPY ./app/pyproject.toml ./app/poetry.lock* /app/
WORKDIR /app/

# Allow installing dev dependencies to run tests
ARG INSTALL_DEV=false
RUN bash -c "if [ $INSTALL_DEV == 'true' ] ; then poetry install --no-root ; else poetry install --no-root --no-dev ; fi"

ARG BACKEND_APP_MODULE=app.main:app
ARG BACKEND_PRE_START_PATH=/app/prestart.sh
ARG BACKEND_PROCESS_MANAGER=gunicorn
ARG BACKEND_WITH_RELOAD=false
ENV APP_MODULE=${BACKEND_APP_MODULE} PRE_START_PATH=${BACKEND_PRE_START_PATH} PROCESS_MANAGER=${BACKEND_PROCESS_MANAGER} WITH_RELOAD=${BACKEND_WITH_RELOAD}
COPY ./app/ /app/
```


When running this I'm getting `Error while finding module specification for 'inboard.start' (ModuleNotFoundError: No module named 'inboard')` and the container exits. Accessing the exited container and inspecting poetry yields

```
root@deb617664e45:/app# whereis poetry
poetry: /opt/pipx/bin/poetry
root@deb617664e45:/app# poetry config --list
cache-dir = "/root/.cache/pypoetry"
experimental.new-installer = true
installer.parallel = true
virtualenvs.create = true
virtualenvs.in-project = null
virtualenvs.path = "{cache-dir}/virtualenvs"  # /root/.cache/pypoetry/virtualenvs
root@deb617664e45:/app# env | grep POETRY
POETRY_VERSION=1.1.11
root@deb617664e45:/app# ls -a /app/.venv
ls: cannot access '/app/.venv': No such file or directory
```

At runtime I can confirm that setting `poetry config virtualenvs.in-project true` makes possible `poetry install` and `python -m inboard.start`.  At build time I can achieve the same result with `ENV POETRY_VIRTUALENVS_IN_PROJECT=true` just before `poetry install`. 

This solves the problem, I think. Thanks!

**Reply by br3ndonland** [reply]
Ok cool! Yep `poetry config virtualenvs.in-project true` will be needed so that Poetry uses the virtualenv. This setting should already be there (inboard will copy the `poetry.toml` in to the container). Could have been a Docker caching issue. I'm glad it works now!
